static/js/util/profile_edit.js

```diff
@@ -191,37 +193,38 @@ export function boundDateField(keySet: string[], label: string, omitDay: boolean
       day: day !== undefined ? day : edit.day
     });
 
-    // these functions return undefined if a month, day, or year is invalid, and converts the value to a number
-    // except if the input value is an empty string, in which case an empty string is returned
-    let validatedDay = 1;
+    let validatedDay = Maybe.of(1);
     if (!omitDay) {
       validatedDay = validateDay(newEdit.day);
-      newEdit.day = validatedDay === undefined ? '' : String(validatedDay);
+      newEdit.day = mstr(validatedDay);
     }
 
     let validatedMonth = validateMonth(newEdit.month);
-    newEdit.month = validatedMonth === undefined ? '' : String(validatedMonth);
+    newEdit.month = mstr(validatedMonth);
 
     let validatedYear = validateYear(newEdit.year);
-    newEdit.year = validatedYear === undefined ? '' : String(validatedYear);
+    newEdit.year = mstr(validatedYear);
 
     // keep text up to date
     _.set(clone, editKeySet, newEdit);
 
-    if ([validatedDay, validatedMonth, validatedYear].includes(undefined)) {
-      // store the edit value and make the formatted date undefined so it doesn't pass validation
+    let padYear = s => _.padStart(s, 4, '0');
+
+    let stringifyDates = () => `${mstr(validatedYear.map(padYear))}-${mstr(validatedMonth)}-${mstr(validatedDay)}`;
+
+    let dateList = [validatedYear, validatedMonth, validatedDay];
+
+    let dateString = S.maybe("", stringifyDates, allJust(dateList));
```

Comment:
If we can change `allJust` to return `Just(dateList)`, we should change `stringifyDates` to take year, month, and day as parameters instead of using the values from the context here